src/main/java/fi/aalto/cs/apluscourses/ui/base/TreeView.java

```diff
@@ -123,12 +122,12 @@ private void update() {
     setSelectionRows(selection);
   }
 
-  public void addNodeAppliedListener(ActionListener listener) {
-    nodeAppliedListeners.add(listener);
+  public void addNodeAppliedListener(@NotNull String key, @NotNull ActionListener listener) {
+    nodeAppliedListeners.put(key, listener);
```

Comment:
This is potentially confusing: `addNodeAppliedListener` suggest that the key is just some String, but then inside `mouseClicked` the key is actually expected to be a class string, not just a generic string.

If the change from a set to a map is necessary, consider changing the first parameter of this function to `Class<?>` and then doing `nodeAppliedListeners.put(key.getName(), listener)` to enforce this contract.